ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -1372,6 +1549,77 @@ private Offset getLastRecordOffset(LogSegment segment) {
     return lastOffset;
   }
 
+  // hardDeletePauseAndResumeOnRestart() helpers
+
+  /**
+   * Utility to test hard delete pause and resume with restart.
+   * @param deleteShutdownFile {@code true} if shutdown file needs to be deleted during reload of index, {@code false}
+   *                                       otherwise
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws StoreException
+   */
+  private void testHardDeletePauseResumeWithRestartTest(boolean deleteShutdownFile)
+      throws InterruptedException, IOException, StoreException {
+    properties.put("store.deleted.message.retention.days", Integer.toString(1));
+    properties.put("store.hard.delete.bytes.per.sec", Integer.toString(Integer.MAX_VALUE / 10));
+    properties.setProperty("store.data.flush.interval.seconds", "3600");
+    // enable daemon thread to run hard deletes
+    properties.put("store.enable.hard.delete", "true");
+    reloadIndex(false);
+    assertTrue("Hard delete is not running", index.hardDeleter.isRunning());
+    // IndexSegment still uses real time so advance time so that it goes 2 days past the real time.
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // give it some time so that hard delete completes one cycle
+    waitUntilExpectedProgress(expectedProgress, 1000);
+    verifyEntriesForHardDeletes(deletedKeys, true);
+
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    advanceTime(HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS);
+    waitUntilExpectedState(Thread.State.WAITING, HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS + 1, 10);
+
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    Set<MockId> idsToDelete = new HashSet<>();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    waitUntilExpectedState(Thread.State.WAITING, HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS + 1, 10);
+    verifyEntriesForHardDeletes(idsToDelete, false);
+
+    reloadIndex(deleteShutdownFile);
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
```

Comment:
isn't most of this code repeated with the `pauseResume()` test? If you had a variable that indicated whether restart was required or not, can this code be reduced?